File: backend/storage/local_file_storage.py
"""Local file-based storage implementation.

Stores JSON files on disk organised as:
    {base_dir}/{workflow_id}/screen_configs/{screen_id}.json
    {base_dir}/{workflow_id}/manifest.json
"""

import json
import os
from pathlib import Path

from pydantic import ValidationError

from models.schemas import ScreenConfig, SimulationManifest
from storage.storage_interface import StorageInterface, NotFoundError, ParseError


class LocalFileStorage(StorageInterface):
    """Stores Screen Configs and Manifests as JSON files on the local filesystem."""

    def __init__(self, base_dir: str = "./data") -> None:
        self.base_dir = Path(base_dir)

    # ── helpers ──────────────────────────────────────────────

    def _configs_dir(self, workflow_id: str) -> Path:
        return self.base_dir / workflow_id / "screen_configs"

    def _config_path(self, workflow_id: str, screen_id: str) -> Path:
        return self._configs_dir(workflow_id) / f"{screen_id}.json"

    def _manifest_path(self, workflow_id: str) -> Path:
        return self.base_dir / workflow_id / "manifest.json"

    # ── Screen Config operations ─────────────────────────────

    def save_screen_config(self, workflow_id: str, config: ScreenConfig) -> None:
        path = self._config_path(workflow_id, config.screen_id)
        path.parent.mkdir(parents=True, exist_ok=True)
        path.write_text(config.model_dump_json(indent=2), encoding="utf-8")

    def load_screen_config(self, workflow_id: str, screen_id: str) -> ScreenConfig:
        path = self._config_path(workflow_id, screen_id)
        if not path.exists():
            raise NotFoundError(
                f"Screen config not found: workflow={workflow_id}, screen={screen_id}"
            )

        raw = path.read_text(encoding="utf-8")
        try:
            data = json.loads(raw)
        except json.JSONDecodeError as exc:
            raise ParseError(
                f"Failed to parse screen config JSON: workflow={workflow_id}, "
                f"screen={screen_id}"
            ) from exc

        try:
            return ScreenConfig.model_validate(data)
        except ValidationError as exc:
            raise ParseError(
                f"Invalid screen config: {exc}"
            ) from exc

    def list_screen_configs(self, workflow_id: str) -> list[ScreenConfig]:
        configs_dir = self._configs_dir(workflow_id)
        if not configs_dir.exists():
            return []

        configs: list[ScreenConfig] = []
        for file_path in sorted(configs_dir.glob("*.json")):
            screen_id = file_path.stem
            try:
                configs.append(self.load_screen_config(workflow_id, screen_id))
            except (NotFoundError, ParseError):
                # skip corrupt / invalid files
                continue
        return configs

    def delete_screen_config(self, workflow_id: str, screen_id: str) -> None:
        path = self._config_path(workflow_id, screen_id)
        if not path.exists():
            raise NotFoundError(
                f"Screen config not found: workflow={workflow_id}, screen={screen_id}"
            )
        os.remove(path)

    # ── Manifest operations ──────────────────────────────────

    def save_manifest(self, workflow_id: str, manifest: SimulationManifest) -> None:
        path = self._manifest_path(workflow_id)
        path.parent.mkdir(parents=True, exist_ok=True)
        path.write_text(manifest.model_dump_json(indent=2), encoding="utf-8")

    def load_manifest(self, workflow_id: str) -> SimulationManifest:
        path = self._manifest_path(workflow_id)
        if not path.exists():
            raise NotFoundError(f"Manifest not found: workflow={workflow_id}")

        raw = path.read_text(encoding="utf-8")
        try:
            data = json.loads(raw)
        except json.JSONDecodeError as exc:
            raise ParseError(
                f"Failed to parse manifest JSON: workflow={workflow_id}"
            ) from exc

        try:
            return SimulationManifest.model_validate(data)
        except ValidationError as exc:
            raise ParseError(f"Invalid manifest: {exc}") from exc

    def manifest_exists(self, workflow_id: str) -> bool:
        return self._manifest_path(workflow_id).exists()

File: frontend/src/storage/LocalFileStorage.ts
import type { ScreenConfig } from "../types/screenConfig";
import type { SimulationManifest } from "../types/manifest";
import type { StorageInterface } from "./StorageInterface";
import { NotFoundError, ParseError } from "./StorageInterface";
import { validateScreenConfig } from "../utils/validateScreenConfig";
import { validateManifest } from "../utils/validateManifest";

/**
 * Key-naming convention in localStorage:
 *   Screen config: `sc:{workflowId}:{screenId}`
 *   Screen config index: `sc-index:{workflowId}` (JSON array of screenIds)
 *   Manifest: `manifest:{workflowId}`
 *
 * This is a frontend-friendly "local" implementation.
 * A teammate will swap this with an API-backed implementation later.
 */
export class LocalFileStorage implements StorageInterface {
  private storage: Storage;

  constructor(storage: Storage = localStorage) {
    this.storage = storage;
  }

  // ── helpers ──────────────────────────────────────────────

  private configKey(workflowId: string, screenId: string): string {
    return `sc:${workflowId}:${screenId}`;
  }

  private indexKey(workflowId: string): string {
    return `sc-index:${workflowId}`;
  }

  private manifestKey(workflowId: string): string {
    return `manifest:${workflowId}`;
  }

  private getScreenIds(workflowId: string): string[] {
    const raw = this.storage.getItem(this.indexKey(workflowId));
    if (!raw) return [];
    try {
      const ids = JSON.parse(raw);
      return Array.isArray(ids) ? ids : [];
    } catch {
      return [];
    }
  }

  private setScreenIds(workflowId: string, ids: string[]): void {
    this.storage.setItem(this.indexKey(workflowId), JSON.stringify(ids));
  }

  // ── Screen Config operations ─────────────────────────────

  async saveScreenConfig(workflowId: string, config: ScreenConfig): Promise<void> {
    this.storage.setItem(
      this.configKey(workflowId, config.screen_id),
      JSON.stringify(config),
    );

    const ids = this.getScreenIds(workflowId);
    if (!ids.includes(config.screen_id)) {
      ids.push(config.screen_id);
      this.setScreenIds(workflowId, ids);
    }
  }

  async loadScreenConfig(workflowId: string, screenId: string): Promise<ScreenConfig> {
    const raw = this.storage.getItem(this.configKey(workflowId, screenId));
    if (raw === null) {
      throw new NotFoundError(
        `Screen config not found: workflow=${workflowId}, screen=${screenId}`,
      );
    }

    let parsed: unknown;
    try {
      parsed = JSON.parse(raw);
    } catch {
      throw new ParseError(
        `Failed to parse screen config JSON: workflow=${workflowId}, screen=${screenId}`,
      );
    }

    const result = validateScreenConfig(parsed);
    if (!result.valid) {
      throw new ParseError(
        `Invalid screen config: ${result.errors.map((e) => e.message).join("; ")}`,
      );
    }

    return parsed as ScreenConfig;
  }

  async listScreenConfigs(workflowId: string): Promise<ScreenConfig[]> {
    const ids = this.getScreenIds(workflowId);
    const configs: ScreenConfig[] = [];

    for (const id of ids) {
      try {
        const config = await this.loadScreenConfig(workflowId, id);
        configs.push(config);
      } catch {
        // skip entries that can't be loaded (deleted / corrupt)
      }
    }

    return configs;
  }

  async deleteScreenConfig(workflowId: string, screenId: string): Promise<void> {
    this.storage.removeItem(this.configKey(workflowId, screenId));

    const ids = this.getScreenIds(workflowId);
    const filtered = ids.filter((id) => id !== screenId);
    this.setScreenIds(workflowId, filtered);
  }

  // ── Manifest operations ──────────────────────────────────

  async saveManifest(workflowId: string, manifest: SimulationManifest): Promise<void> {
    this.storage.setItem(this.manifestKey(workflowId), JSON.stringify(manifest));
  }

  async loadManifest(workflowId: string): Promise<SimulationManifest> {
    const raw = this.storage.getItem(this.manifestKey(workflowId));
    if (raw === null) {
      throw new NotFoundError(`Manifest not found: workflow=${workflowId}`);
    }

    let parsed: unknown;
    try {
      parsed = JSON.parse(raw);
    } catch {
      throw new ParseError(`Failed to parse manifest JSON: workflow=${workflowId}`);
    }

    const result = validateManifest(parsed);
    if (!result.valid) {
      throw new ParseError(
        `Invalid manifest: ${result.errors.map((e) => e.message).join("; ")}`,
      );
    }

    return parsed as SimulationManifest;
  }

  async manifestExists(workflowId: string): Promise<boolean> {
    return this.storage.getItem(this.manifestKey(workflowId)) !== null;
  }
}
